core/capstanignore.go

```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright (C) 2017 XLAB, Ltd.
+ *
+ * This work is open source software, licensed under the terms of the
+ * BSD license as described in the LICENSE file in the top-level directory.
+ */
+
+package core
+
+import (
+	"bufio"
+	"fmt"
+	"os"
+	"regexp"
+	"strings"
+)
+
+type Capstanignore interface {
+	LoadFile(path string) error
+	AddPattern(pattern string) error
+	PrintPatterns()
+	IsIgnored(path string) bool
+}
+
+var CAPSTANIGNORE_ALWAYS []string = []string{"/meta/*", "/mpm-pkg/*", "/.git/*"}
+
+// CapstanignoreInit creates a new capstanignore struct that is
+// used when deciding whether a file should be included in unikernel
+// or not. You can provide `path` to the .capstanignore file to load
+// it or pass empty string "" if you have none. Note that once having
+// capstanignore struct you can load as many files as you want (using
+// .LoadFile function) or manually add as many patterns as you like
+// (using .AddPattern function).
+func CapstanignoreInit(path string, verbose bool) Capstanignore {
+	c := capstanignore{verbose: verbose}
+
+	// Load capstanignore file if path was given.
+	if path != "" {
+		if err := c.LoadFile(path); verbose && err != nil {
+			fmt.Println("WARN: failed to load .capstanignore file:", err)
+		}
+	}
+
+	// Always ignore some common paths.
+	for _, pattern := range CAPSTANIGNORE_ALWAYS {
+		c.AddPattern(pattern)
+	}
+	return &c
+}
+
+type capstanignore struct {
+	ignored  []string         // list of all ignored patterns
+	ignoredC []*regexp.Regexp // list of compiled patterns
+	verbose  bool
+}
+
+// LoadFile attempts to parse .capstanignore file on given path.
+// If success, it remembers all patterns and closes file.
+func (c *capstanignore) LoadFile(path string) error {
+	if file, err := os.Open(path); err == nil {
+		defer file.Close()
+
+		scanner := bufio.NewScanner(file)
+		for scanner.Scan() {
+			line := strings.Trim(scanner.Text(), " ")
+			if len(line) == 0 || strings.HasPrefix(line, "#") {
+				continue
+			}
+			if errPattern := c.AddPattern(line); errPattern != nil {
+				return errPattern
+			}
+		}
+	}
+	return nil
+}
+
+// AddPattern adds a pattern to be ignored.
+func (c *capstanignore) AddPattern(pattern string) error {
+	safePattern := transformCapstanignoreToRegex(pattern)
+	if compiled, err := regexp.Compile(safePattern); err == nil {
+		c.ignored = append(c.ignored, pattern)
+		c.ignoredC = append(c.ignoredC, compiled)
+	} else {
+		return err
+	}
+
+	// also ignore folder when all its content is ignored
+	for strings.HasSuffix(pattern, "/*") {
+		pattern = strings.TrimSuffix(pattern, "/*")
+		safePattern = transformCapstanignoreToRegex(pattern)
+		compiled := regexp.MustCompile(safePattern)
+		c.ignoredC = append(c.ignoredC, compiled)
+	}
+
+	return nil
+}
+
+func (c *capstanignore) IsIgnored(path string) bool {
+	for _, pattern := range c.ignoredC {
+		if pattern.MatchString(path) {
```

Comment:
Just out of curiosity: would it be faster to have one large regex consisting of all ignore patterns instead of this loop? Namely, `(patt1|patt2|...)`